### PR TITLE
Weak cache

### PR DESCRIPTION
**Pull request by AlexUstinov**
- access MappingInfo property by reference 
    ^ changed it to copying value, because `Seqment` is 2-field readonly struct - if we use both fields it is more efficient to copy entire struct than dereferencing a few times.

- allocate internal dictionary for the WeakCache on demand

- Use `Dictionary.Remove(key, out var removeElement)` overload to save one search operation in dictionary
- Use `Dictionary(capacity = NoGcCount)` constructor to allocate enough space from beginning

### CODE REVIEW

File: Orm/Xtensive.Orm/Caching/WeakCache.cs [hunk 1]
```diff
@@ -229,26 +217,19 @@ public virtual void CollectGarbage()
...
-    }
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
     /// <inheritdoc/>
     public virtual IEnumerator<TItem> GetEnumerator()
     {
-      foreach (var pair in items) {
-        var item = ExtractTarget(pair.Value);
-        if (item!=null)
+      foreach (var pair in items ?? Enumerable.Empty<KeyValuePair<TKey, GCHandle>>()) {
```
Comment (botinko) [inline Orm/Xtensive.Orm/Caching/WeakCache.cs:225]: `if (items == null) return Enumerable.Empty<KeyValuePair<TKey, GCHandle>>` ?

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Caching/WeakCache.cs]: Refactored slightly other way, becuuse it is impossible to use both: `return Collection` & `yield return item` in the sam eiterator method

File: Orm/Xtensive.Orm/Caching/WeakCache.cs [hunk 2]
```diff
@@ -220,13 +220,11 @@ public virtual void CollectGarbage()
...
 
     /// <inheritdoc/>
-    public virtual IEnumerator<TItem> GetEnumerator()
-    {
-      foreach (var pair in items ?? Enumerable.Empty<KeyValuePair<TKey, GCHandle>>()) {
-        if (ExtractTarget(pair.Value) is TItem item)
-          yield return item;
-      }
-    }
+    public virtual IEnumerator<TItem> GetEnumerator() =>
```
Comment (botinko) [inline Orm/Xtensive.Orm/Caching/WeakCache.cs:223]: I just made a bench. Previous version was faster.

Comment (botinko) [reply Orm/Xtensive.Orm/Caching/WeakCache.cs:223]:
```

Comment (botinko) [reply Orm/Xtensive.Orm/Caching/WeakCache.cs:223]: Let's stay with previous version.

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Caching/WeakCache.cs]: reverted

**Review by SergeiPavlov — APPROVED** [review]

File: Orm/Xtensive.Orm/Caching/WeakCache.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Caching/WeakCache.cs:225]: reverted back because of performance measurement

**Review by botinko — APPROVED** [review]

**Pull request merged**